File: app/src/main/java/i/am/shiro/amai/adapter/DownloadJobAdapter.java
package i.am.shiro.amai.adapter;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import i.am.shiro.amai.R;
import i.am.shiro.amai.model.DownloadJob;

public class DownloadJobAdapter extends ListAdapter<DownloadJob, DownloadJobAdapter.ViewHolder> {

    public DownloadJobAdapter() {
        super(new DiffCalback());
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(R.layout.item_queue_generic, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        DownloadJob job = getItem(position);
        holder.bindData(job);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView titleText;

        private final TextView statusText;

        ViewHolder(View itemView) {
            super(itemView);
            titleText = itemView.findViewById(R.id.titleText);
            statusText = itemView.findViewById(R.id.statusText);
        }

        void bindData(DownloadJob job) {
            titleText.setText(job.getTitle());
            statusText.setText(job.getStatusString());
        }
    }

    private static class DiffCalback extends DiffUtil.ItemCallback<DownloadJob> {

        @Override
        public boolean areItemsTheSame(@NonNull DownloadJob oldItem, @NonNull DownloadJob newItem) {
            return oldItem.getBookId() == newItem.getBookId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull DownloadJob oldItem, @NonNull DownloadJob newItem) {
            return oldItem.getStatus() == newItem.getStatus();
        }
    }
}

File: app/src/main/java/i/am/shiro/amai/adapter/DownloadJobAdapter.kt
package i.am.shiro.amai.adapter

import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import i.am.shiro.amai.DownloadStatus
import i.am.shiro.amai.data.model.Download
import i.am.shiro.amai.databinding.ItemDownloadDoneBinding
import i.am.shiro.amai.databinding.ItemDownloadDownloadingBinding
import i.am.shiro.amai.databinding.ItemDownloadFailedBinding
import i.am.shiro.amai.databinding.ItemDownloadGenericBinding
import i.am.shiro.amai.util.inflateChild

class DownloadJobAdapter(
    private val onDismiss: (Download) -> Unit,
    private val onCancel: (Download) -> Unit,
    private val onRetry: (Download) -> Unit,
    private val onPause: (Download) -> Unit,
    private val onResume: (Download) -> Unit
) : ListAdapter<Download, DownloadJobAdapter.ViewHolder>(DiffCallback()) {

    override fun getItemViewType(position: Int) = getItem(position).status.ordinal

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return when (DownloadStatus.values()[viewType]) {
            DownloadStatus.DONE -> {
                DoneViewHolder(parent.inflateChild(ItemDownloadDoneBinding::inflate))
            }
            DownloadStatus.FAILED -> {
                FailedViewHolder(parent.inflateChild(ItemDownloadFailedBinding::inflate))
            }
            DownloadStatus.RUNNING -> {
                DownloadingViewHolder(parent.inflateChild(ItemDownloadDownloadingBinding::inflate))
            }
            else -> {
                GenericViewHolder(parent.inflateChild(ItemDownloadGenericBinding::inflate))
            }
        }
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val job = getItem(position)
        holder.bindData(job)
    }

    abstract class ViewHolder(binding: ViewBinding) : RecyclerView.ViewHolder(binding.root) {

        internal abstract fun bindData(job: Download)
    }

    internal class GenericViewHolder(private val binding: ItemDownloadGenericBinding) :
        ViewHolder(binding) {

        override fun bindData(job: Download) {
            with(binding) {
                titleText.text = job.title
                statusText.text = job.status.toString()
            }
        }
    }

    internal inner class DoneViewHolder(private val binding: ItemDownloadDoneBinding) :
        ViewHolder(binding) {

        override fun bindData(job: Download) {
            with(binding) {
                titleText.text = job.title
                dismissButton.setOnClickListener { onDismiss(job) }
            }
        }
    }

    internal inner class FailedViewHolder(private val binding: ItemDownloadFailedBinding) :
        ViewHolder(binding) {

        override fun bindData(job: Download) {
            with(binding) {
                binding.titleText.text = job.title
                cancelButton.setOnClickListener { onCancel(job) }
                retryButton.setOnClickListener { onRetry(job) }
            }
        }
    }

    internal inner class DownloadingViewHolder(private val binding: ItemDownloadDownloadingBinding) :
        ViewHolder(binding) {

        override fun bindData(job: Download) {
            with(binding) {
                binding.titleText.text = job.title
                progressBar.max = job.progressMax
                progressBar.progress = job.progress
                cancelButton.setOnClickListener { onCancel(job) }
                pauseButton.setOnClickListener { onPause(job) }
            }
        }
    }

    private class DiffCallback : DiffUtil.ItemCallback<Download>() {

        override fun areItemsTheSame(oldItem: Download, newItem: Download) =
            oldItem.bookId == newItem.bookId

        override fun areContentsTheSame(oldItem: Download, newItem: Download) =
            oldItem.status == newItem.status && oldItem.progress == newItem.progress
    }
}
